bliss/models/lens_encoder.py

```diff
@@ -0,0 +1,356 @@
+            elif param_idx in [3, 4]: # e1, e2 (3, 4) need to invert: torch.logit((e1 + 1) / 2): TODO: change to angle to constrain relation
+                param = torch.sigmoid(param) * 2 - 1
+            lens_params[...,param_idx + self.latent_dim_split] = param
+        
+        return lens_params.to(device=var_dist_params.device)
+
+    def training_step(self, batch, batch_idx):
+        """Pytorch lightning training step."""
+        batch_size = len(batch["images"])
+        loss = self._get_loss(batch)
+        self.log("train/loss", loss, batch_size=batch_size)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        """Pytorch lightning validation step."""
+        batch_size = len(batch["images"])
+        pred = self._get_loss(batch)
+        self.log("val/loss", pred["loss"], batch_size=batch_size)
+        pred_out = {f"pred_{k}": v for k, v in pred.items()}
+        return {**batch, **pred_out}
+
+    def validation_epoch_end(self, outputs):
+        """Pytorch lightning method run at end of validation epoch."""
+        # put all outputs together into a single batch
+        batch = {}
+        for b in outputs:
+            for k, v in b.items():
+                if len(v.shape) > 0:
+                    curr_val = batch.get(k, torch.tensor([], device=v.device))
+                    batch[k] = torch.cat([curr_val, v])
+        if self.n_bands == 1:
+            self._make_plots(batch)
+
+    def _get_loss(self, batch):
+        images: Tensor = batch["images"]
+        background: Tensor = batch["background"]
+        tile_catalog = TileCatalog(
+            self.tile_slen, {k: v for k, v in batch.items() if k not in {"images", "background"}}
+        )
+
+        image_ptiles = get_images_in_tiles(
+            torch.cat((images, background), dim=1),
+            self.tile_slen,
+            self.ptile_slen,
+        )
+        image_ptiles = rearrange(image_ptiles, "n nth ntw b h w -> (n nth ntw) b h w")
+        locs = rearrange(tile_catalog.locs, "n nth ntw ns hw -> 1 (n nth ntw) ns hw")
+        lens_pred = self.encode(image_ptiles, locs)
+        lens_pred = rearrange(
+            lens_pred,
+            "ns (n nth ntw) ms d -> (ns n) nth ntw ms d",
+            ns=1,
+            nth=tile_catalog.n_tiles_h,
+            ntw=tile_catalog.n_tiles_w,
+        )
+
+        lensed_galaxy_pred, pure_lens_pred = lens_pred[...,:2 * self.latent_dim_split], lens_pred[...,2 * self.latent_dim_split:]
+        lensed_galaxy_params, pure_lens_params = batch["lens_params"][...,:self.latent_dim_split], batch["lens_params"][...,self.latent_dim_split:]
+
+        loss_lens = self._get_params_logprob(
+            batch["lensed_galaxy_bools"],
+            pure_lens_params,
+            pure_lens_pred,
+        ) 
+        
+        loss_lensed_galaxy = get_galsim_params_nll(
+            batch["lensed_galaxy_bools"],
+            lensed_galaxy_params,
+            lensed_galaxy_pred,
+        )
+
```

Comment:
Is the lens encoder in charge of both predicting the lens params and the galaxy params at the center of the lens? I wonder if the galaxy encoder could be in charge of the latter and the lens encoder of the former. I wonder if that could improve performance too.